AC-329 making closed-captions draggable

clrux opened the pull request:
# [AC-329](https://openedx.atlassian.net/browse/AC-329)

This work completes the video player closed captioning by making the captions draggable. Learners can now move the closed captions anywhere within the video player. It is the continuation of [AC-188](https://openedx.atlassian.net/browse/AC-188).

Additionally, it fixes references to A Font Garde and Modernizr, which are required to use the accessible icon fonts that were included in the video player updates a few months back.

## Sandbox 

https://clrux-ac-329.sandbox.edx.org
https://studio-clrux-ac-329.sandbox.edx.org

## Reviewers

- [x] @clytwynec (Team reviewer)
- [x] @cahrens (T&L and second pair of eyes)

clrux commented:
jenkins run js

cahrens commented:
Make sure you look at  #11320 and include draggabilly to the list of files that needs special handling. Then on a sandbox, verify that the file does not appear without an MD5 hash in the name.

clrux commented:
@cahrens Thanks. I'm trying to make sense of it, and I've added it to the overrides section. Do I keep it in  all of the other locations as well, or only use this one?

cahrens commented:
Keep it in the other locations also.

clrux commented:
@cahrens 
![dog-meme-i-have-no-idea-what-im-doing-03](https://cloud.githubusercontent.com/assets/2112024/12857789/1c0328d6-cc1a-11e5-89fb-50ba8764a36c.jpg)

clrux commented:
jenkins run all

clrux commented:
@cahrens No idea what keeps going on here. Got time to help?

cahrens commented on lms/envs/common.py:
```diff
@@ -1239,6 +1239,7 @@
     'pattern-library/js/modernizr.custom.js',
     'pattern-library/js/afontgarde.js',
     'pattern-library/js/edx-icons.js'
+    'js/vendor/draggabilly.pkgd.js',
```
Don't put draggabilly in main vendor files. It's not used very much.

cahrens commented on cms/static/coffee/spec/main_squire.coffee:
```diff
@@ -176,6 +176,9 @@ requirejs.config({
         "coffee/src/ajax_prefix": {
             deps: ["jquery"]
+        },
+        "draggabilly": {
```
Probably don't actually need this here, though if the squire and non-squire versions are meant to be in sync, it's fine to leave it.

cahrens commented:
Try removing main_vendor_js from main_vendor_js and see if that helps (t shouldn't be there in any case).

clrux commented:
@cahrens I don't understand that last comment. Can you clarify?

cahrens commented:
@clrux I updated my comment. Sorry for the typo.

clrux commented:
@cahrens Removed, but there are an insane number of failing tests. Or so Jenkins says.

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -1138,6 +1140,17 @@
+
+                if (typeof Draggabilly === "function" && captions) {
+                    draggable = new Draggabilly(captions, { containment: true });
+                } else {
```
It's unclear why you can assume that the reason the if statement is False is that draggabilly isn't there.  If the captions element is always present, then it shouldn't need to be part of that check.  If it isn't, then this error message could be reporting an incorrect reason for the missing functionality.

clytwynec commented:
@clrux Can you please add some tests for this new functionality?

clrux commented:
@clytwynec Do I write tests to see if the captions are draggable? Is that something that can be tested?

clrux commented:
jenkins run bokchoy
jenkins run lettuce
jenkins run quality

clrux commented:
jenkins run all

cahrens commented on cms/static/cms/js/require-config.js:
```diff
@@ -300,6 +300,9 @@
                        "tags-annotator", "flagging-annotator", "grouping-annotator", "diacritic-annotator",
                        "openseadragon", "jquery-Watch", "catch", "handlebars", "URI"]
+            },
+            "draggabilly": {
```
I don't think you need this in CMS (the shim part) because CMS uses RequireJS throughout.

cahrens commented on lms/static/js/spec/main.js:
```diff
@@ -66,6 +65,7 @@
             '_split': 'js/split',
             'mathjax_delay_renderer': 'coffee/src/mathjax_delay_renderer',
             'MathJaxProcessor': 'coffee/src/customwmd',
+            'draggabilly': 'js/vendor/draggabilly.pkgd',
```
dragabbillly is above with the correct path, remove this

cahrens commented on common/lib/xmodule/xmodule/js/js_test.yml:
```diff
@@ -59,6 +59,7 @@ lib_paths:
     - public/js/split_test_staff.js
     - common_static/js/src/accessibility_tools.js
     - common_static/js/vendor/moment.min.js
+    - common_static/js/vendor/draggabilly.pkgd.js
```
Also add draggabilly to the RequireJS config in spec/main_requirejs.js (which is only used for xmodule tests).

clrux commented:
jenkins run bokchoy

clytwynec commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -1184,6 +1185,39 @@ def test_quality_button_works_correctly(self):
+        captions = self.video.q(css=captions_container).results[0]
+
+        action = ActionChains(self.browser)
+        action.drag_and_drop_by_offset(captions, 15, 185).perform()
```
It  seems like there should be some other assert statement in here.  A typical pattern would be something along the lines of getting the initial position of the item, performing the `drag_and_drop_by_offset`,  and then waiting for the position to be changed by the specified offset.

clytwynec commented:
@clrux  I had one comment on the tests. Otherwise, the code looks good to me.

I did notice when I was looking at the sandbox that there is a new js error in on the courseware pages
(`Uncaught TypeError: define is not a function`).  Maybe you've already fixed that and just haven't updated the sandbox recently?

clrux commented:
@clytwynec I've modified the test to actually drag the captions and check to ensure they were in fact dragged. This passes.

@cahrens While tests pass and everything works, both the Draggabilly script and the MomentJS script are appearing "unhashed". What am I missing?

cahrens commented on lms/envs/common.py:
```diff
@@ -1248,9 +1248,9 @@
-    'js/vendor/afontgarde/modernizr.fontface-generatedcontent.js',
-    'js/vendor/afontgarde/afontgarde.js',
-    'js/vendor/afontgarde/edx-icons.js'
+    'pattern-library/js/modernizr.custom.js',
```
@clrux You need to verify with @andy-armstrong that you can make these changes. I believe this was tried once before and did not work.

cahrens commented on lms/envs/common.py:
```diff
@@ -1633,7 +1633,8 @@
     'js/bookmarks/views/bookmark_button.js',
-    'js/views/message_banner.js'
+    'js/views/message_banner.js',
+    'js/vendor/draggabilly.pkgd.js'
```
@clrux This path may not be correct (it may need a prefix since draggabilly lives in common, whereas the other two files are LMS files). This is the configuration that should prevent the uncached version of the file from appearing. I recommend that you work with @muhammad-ammar to sort out why you are still seeing an uncached file.

It is expected that moment is uncached-- there is an outstanding bug to fix it (presumably in the same way that you are registering draggabilly).

clrux commented:
jenkins run quality

andy-armstrong commented on lms/envs/common.py:
```diff
@@ -1248,9 +1248,9 @@
-    'js/vendor/afontgarde/modernizr.fontface-generatedcontent.js',
-    'js/vendor/afontgarde/afontgarde.js',
-    'js/vendor/afontgarde/edx-icons.js'
+    # 'pattern-library/js/modernizr.custom.js',
```
I agree that we shouldn't bundle these. Let's only load them via RequireJS, either through an optimized RequireJS factory bundle, or dynamically if necessary (such as with the video player). Go ahead and remove these commented out lines.

andy-armstrong commented on lms/envs/common.py:
```diff
@@ -1637,7 +1637,8 @@
     'js/courseware/course_home_events.js',
-    'js/courseware/toggle_element_visibility.js'
+    'js/courseware/toggle_element_visibility.js',
+    'js/vendor/draggabilly.pkgd.js',
```
As we discussed, you can add the Pattern Library JS files here so that they will be loaded correctly by the video player.

andy-armstrong commented on cms/static/cms/js/require-config.js:
```diff
@@ -63,10 +63,13 @@
             "URI": "js/vendor/URI.min",
             "ieshim": "js/src/ie_shim",
             "tooltip_manager": "js/src/tooltip_manager",
+            "modernizr": "js/vendor/afontgarde/modernizr.fontface-generatedcontent",
```
This looks good for now. Very shortly I'll be removing these files from the repo, so then I'll update these paths to point to the installed Pattern Library versions instead, but we shouldn't hold up this PR for that.

BTW, do you know the difference between the ``modernizr`` file in the Pattern Library and this "generated content" version. Do you see a problem with replacing one with the other?

clrux replied:
If the one with the library has both "fontface" and "generated-content" support, it should work for A Font Garde as well and we would be able to use it instead.

andy-armstrong commented on cms/static/cms/js/require-config.js:
```diff
@@ -249,6 +251,21 @@
+                deps: ["jquery", "modernizr"],
+                exports: "AFontGarde"
+            },
+            "edxicons": {
```
A shim shouldn't be necessary as ``edx-icons`` is implemented using RequireJS. I'm surprised that this works, in fact.

clrux replied:
Removing these made it not work, actually.

andy-armstrong commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -207,7 +210,19 @@ requirejs.config({
+            deps: ["jquery", "modernizr"],
+            exports: "AFontGarde"
+        },
+        "edxicons": {
```
Same deal about not shimming edxicons.

clrux replied:
Removing these made it not work, actually.

andy-armstrong commented on cms/static/cms/js/require-config.js:
```diff
@@ -63,10 +63,13 @@
             "tooltip_manager": "js/src/tooltip_manager",
+            "modernizr": "js/vendor/afontgarde/modernizr.fontface-generatedcontent",
+            "afontgarde": "js/vendor/afontgarde/afontgarde",
+            "edxicons": "js/vendor/afontgarde/edx-icons",
```
It is rather strange to have edx-icons live in vendor, when it is not a vendor library. I'd prefer to load it directly from the Pattern Library but that will require my other changes.

clrux replied:
Mind updating this when you get to your piece of work?

clrux commented:
Finally! @clytwynec I've got the captions draggable and the file hashed. Would you take another stroll through?

@cahrens or @andy-armstrong Heads-up here. If you see anything glaring, please let me know.

Sandbox is updated.

cahrens commented on lms/envs/common.py:
```diff
@@ -1247,7 +1247,8 @@
     'js/vendor/URI.min.js',
-    'js/vendor/backbone-min.js'
+    'js/vendor/backbone-min.js',
+    'js/vendor/afontgarde/modernizr.fontface-generatedcontent.js',
```
Why is it necessary to put this modernizr library in base_vendor? When will it be used?

We generally want to load as little as possible in base_vendor.

clrux replied:
This modernizr file is used by the edxicons and the new accessible icon fonts. While they're not yet implemented throughout the platform, this paves the way and allows us to start converting them as we need. I added it to the global list because it needs to be available on every page in the platform.

Do you have another suggestion for global scripts like this?

clrux replied:
@cahrens Did you have thoughts on this? @andy-armstrong Do you know if the modernizr package that's in the platform already includes "fontface" and "generatedcontent"? If so, we can omit this one entirely.

andy-armstrong replied:
@clrux The modernizr in the platform is just the copy from the Pattern Library. I believe it is the full version rather than a limited subset, but I don't know for sure. 

It would be great to get rid of these checked in files, and instead replace them with the Pattern Library versions. Maybe that should be a separate pull request.

clrux replied:
Yes. I need these files now and since the UXPL isn't fully integrated, that will need to be a separate pull request. I will try removing this instance of Modernizr though and see if the icons (used in the video player) still work as expected.

clrux replied:
OK after removing this instance of Modernizr, the icon fonts don't have their accessible fallbacks. Since this PR needs to get in and done, I'm going to include this version for now. Whenever the UXPL is integrated into the platform, we can use that version of Modernizr instead of this one.

cahrens commented on lms/static/lms/js/build.js:
```diff
@@ -33,7 +33,8 @@
             'support/js/enrollment_factory',
-            'js/bookmarks/bookmarks_factory'
+            'js/bookmarks/bookmarks_factory',
+            'js/vendor/draggabilly'
```
A vendor file should not be in the optimizer factory list.

clrux commented:
@andy-armstrong Here's a summary of this latest commit:

* updated references to AFontGarde, Modernizr, and edx-icons to those in the edx-pattern-library
* removed the old ones from vendor
* updated the build of modernizr as i was getting an error from the old one
* corrected some RequireJS dependencies that was causing Modernizr and AFontGarde to load twice

clrux commented:
jenkins run a11y
jenkins run bokchoy
jenkins run lettuce

clrux commented:
jenkins run a11y
jenkins run bokchoy
jenkins run lettuce

clrux commented:
jenkins run a11y

clrux commented:
jenkins run a11y
jenkins run bokchoy

clrux commented:
It seems that the version of Modernizr coming from the UXPL is throwing an error. I've updated it locally, but the real fix won't happen until the updated version in the UXPL is merged in.

Furthermore, when testing the drag and drop capabilities, Firefox and Chrome seem to treat the coordinates differently. Firefox drags the container correctly, but Chrome doesn't. In fact, Chrome moves it out of bounds no matter what x/y value is specified.

clrux commented:
jenkins run all

clrux commented:
This work is on hold until https://github.com/edx/ux-pattern-library/pull/312 is merged which contains the fixed version of Modernizr.

clrux commented:
@andy-armstrong This is nearly there. The last piece of this isn't actually related to the dragging functionality at all, rather the new AFontGarde icon fonts, specifically the RequireJS dependencies. When previewing this locally using `paver run_all_servers` Studio correctly loads `edx-icons.js` and adds the `fontawesome` class, however LMS gives me an error saying, `define is not a function`. I'm wondering if it's an implementation difference?

andy-armstrong commented:
@clrux You are right that it is an implementation difference. Studio uses RequireJS to load everything, so you are able to load edx-icons everywhere. LMS doesn't, so you are not able to pre-load edx-icons. As it stands, the only way to get this to work is to have each factory file that needs edx-icons to declare it as a dependency.

Your other option is to change edx-icons so that it can be loaded either with or without RequireJS. If RequireJS is not available, then it would just assume that AFontGarde was globally available. You'll need logic something like this that we added to the UI Toolkit:

https://github.com/edx/edx-ui-toolkit/blob/master/src/js/utils/string-utils.js#L65

We can discuss this tomorrow if you want to explore the options. Sorry this is so painful!

clrux commented:
@andy-armstrong So I'd like to have the edx-icons and AFontGarde available globally so we can start updating icons whenever possible. Are you suggesting we need to do this conditional code in every file that may use icons? Would it be better to add this to the UXPL `edx-icons

clytwynec commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -530,8 +535,8 @@
                                 self.subtitlesEl.find('.subtitles-menu')
-                                    .text(gettext('Transcript will be displayed when you start playing the video.')) // jshint ignore: line
-                                        .wrapInner('<li></li>');
+                                    .text(window.gettext('Transcript will be displayed when you start playing the video.')) // jshint ignore: line
```
Why does this need `window.` now but it didn't before?

clrux replied:
Hm, I thought I changed all of those back. I'm going to remove the `windows` in this use.

clrux commented:
@cahrens or @andy-armstrong do you think one of you would be able to re-review this early next week?

clytwynec commented:
👍  from me.

clrux commented:
jenkins run coveralls

clrux commented:
jenkins run coverage

clrux commented:
jenkins run python

clrux commented:
jenkins run bokchoy

cahrens commented on cms/static/cms/js/require-config.js:
```diff
@@ -241,14 +244,26 @@
+            },
+            "draggabilly": {
+                deps: ["jquery"],
+                exports: "Draggabilly"
```
You should not need the exports here. Draggabilly is RequireJS aware, so if Require is present, it will not export anything.

The "exports" part is only for libraries that don't use RequireJS, and instead add their library on the global window object.

cahrens commented on cms/static/cms/js/require-config.js:
```diff
@@ -241,14 +244,26 @@
+            "afontgarde": {
+                exports: "AFontGarde"
+            },
+            "edxicons": {
```
EdxIcons is also RequireJS aware. You can remove this whole section about edxicons.

https://github.com/edx/ux-pattern-library/blob/master/pattern-library/js/edx-icons.js

cahrens commented on cms/static/coffee/spec/main.coffee:
```diff
@@ -206,6 +209,19 @@ requirejs.config({
+        "afontgarde": {
+            exports: "AFontGarde"
+        },
+        "edxicons": {
```
See same notes as above.

cahrens commented on cms/static/coffee/spec/main_squire.coffee:
```diff
@@ -176,6 +179,15 @@ requirejs.config({
+        "afontgarde": {
+            exports: "AFontGarde"
+        },
+        "edxicons": {
```
Same comment as above.

cahrens commented on cms/static/js_test_squire.yml:
```diff
@@ -75,6 +75,7 @@ src_paths:
     - common/js
     - edx-pattern-library/js
     - edx-ui-toolkit/js
+    - js/vendor/afontgarde
```
This isn't even the correct path, is it? If it's not needed here, just don't add it.

cahrens commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -98,7 +98,7 @@ class HtmlModuleMixin(HtmlBlock, XModule):
             resource_string(__name__, 'js/src/collapsible.js'),
             resource_string(__name__, 'js/src/html/imageModal.js'),
-            resource_string(__name__, 'js/common_static/js/vendor/draggabilly.pkgd.js'),
+            resource_string(__name__, 'js/common_static/js/vendor/draggabilly.js'),
```
I find the usage of draggabilly by the HTML module to be very mysterious. @clrux, do you know how the HTML module uses it? Have you verified that the functionality still works (if it ever did)?

clrux replied:
Good question. I changed the name of the file, then did a find/replace to pick up any references. Other than updating the filename, I assume things with this module would work as they did. This particular module is unrelated to this scope of work. Thoughts?

cahrens replied:
OK, you don't have to sleuth. But I seriously doubt that it is needed by the HTML component (and question whether or not it would actually work).

cahrens commented on common/lib/xmodule/xmodule/js/spec/main_requirejs.js:
```diff
@@ -1,10 +1,27 @@
+        "afontgarde": {
+            exports: "AFontGarde"
+        },
+        "edxicons": {
```
See notes above.

cahrens commented on common/static/common/js/spec/main_requirejs.js:
```diff
@@ -149,6 +153,19 @@
+            "afontgarde": {
+                exports: "AFontGarde"
+            },
+            "edxicons": {
```
See notes above.

cahrens commented on common/test/acceptance/tests/video/test_video_module.py:
```diff
@@ -1189,6 +1191,42 @@ def test_quality_button_works_correctly(self):
+        captions_start = captions.location
+        action.drag_and_drop_by_offset(captions, 0, -15).perform()
+        captions_end = captions.location
+        self.assertEqual(
```
Thanks for adding an acceptance test!

cahrens commented on lms/static/js/spec/main.js:
```diff
@@ -625,8 +628,19 @@
                 ],
                 exports: 'DiscussionSpecHelper'
+            },
+            'draggabilly':  {
```
See notes above.

cahrens commented on package.json:
```diff
@@ -2,8 +2,9 @@
   "name": "edx",
   "version": "0.1.0",
   "dependencies": {
+    "backbone": "1.3.3",
```
Don't put backbone here yet! We are not ready to install this via NPM (the Lahore team is working on a Backbone upgrade).

cahrens commented:
Don't check draggabilly into lms/vendor. There is already a copy of it in common/static/js/vendor, and that is the correct location for vendor files (we wish to remove everything in the lms vendor directory).

cahrens commented:
Oops, hit wrong button when I posted the last comment.

cahrens commented on lms/static/lms/js/require-config.js:
```diff
@@ -200,6 +204,19 @@
+            "afontgarde": {
+                exports: "AFontGarde"
+            },
+            "edxicons": {
```
See notes above.

The tricky case will be to make sure that in the LMS, the optimizer is able to handle the way that edxicons and draggabilly are checking for the presence of RequireJS. It has to be a very specific syntax, or else the optimizer is not able to modify the check to work with our namespaced version of RequireJS (specific to the LMS). If this doesn't work properly, you will see those vendor files appearing individually (and most likely not cached) on the sandbox.

If this is an issue, @andy-armstrong or I can help.

clrux replied:
Dragging works with these included, but not without them.

cahrens commented on lms/static/lms/js/build.js:
```diff
@@ -34,7 +34,8 @@
             'js/bookmarks/bookmarks_factory',
-            'js/learner_dashboard/program_list_factory'
+            'js/learner_dashboard/program_list_factory',
+            'js/vendor/draggabilly',
```
Draggabilly should not be listed here. This list is for high-level pages, where we want all dependencies of those pages to be bundled into a single file.

cahrens commented on cms/static/js_test.yml:
```diff
@@ -56,7 +56,7 @@ lib_paths:
     - xmodule_js/src/xmodule.js
     - xmodule_js/common_static/js/test/i18n.js
-    - xmodule_js/common_static/js/vendor/draggabilly.pkgd.js
+    - xmodule_js/common_static/js/vendor/draggabilly.js
```
I don't understand this change because the name of the file in common/static/js/vendor is draggabilly.pkgd.js.

clrux commented on lms/static/lms/js/require-config.js:
```diff
@@ -210,6 +210,13 @@
+            "edxicons": {
+                exports: "edxicons"
+            },
+            "draggabilly": {
```
Add a comment here stating that:

Because Draggabilly is being used by video code, the namespaced version of require is not being recognized. Therefore the library is being added to the global namespace instead of being registered in require.

clrux commented on cms/static/cms/js/require-config.js:
```diff
@@ -257,13 +257,6 @@
-                exports: "edxicons"
-            },
-            "draggabilly": {
-                deps: ["jquery"],
```
Restore the "deps" part if Draggabilly does require JQuery.

clrux commented:
@cahrens Sandbox has been rebuilt with the latest changes.

cahrens commented:
Unfortunately 2 of the edx-pattern-library JS files are not cached (see the lack of a hash in the filename). Interestingly, afontgarde IS cached.

![image](https://cloud.githubusercontent.com/assets/484484/14529347/2123c292-0222-11e6-972c-12ec0c410e5b.png)

clrux commented:
Hm, do they need to also be included in the LMS RequireJS file where I added Draggabilly?

cahrens commented:
@clrux do you know if the video player can appear anywhere outside of the courseware? If so, have you verified that Draggabilly is working in those locations?

cahrens commented on lms/envs/common.py:
```diff
@@ -1261,17 +1261,15 @@
     'js/vendor/URI.min.js',
-    'js/vendor/backbone-min.js'
+    'js/vendor/backbone-min.js',
+    'edx-pattern-library/js/modernizr-custom.js',
```
Because modernizr-custom is listed here, it is being bundled into the single vendor file (on the courseware, this is lms-main_vendor. However, it is also being downloaded as a separate file. We don't want this situation.

Why did you decide to put it in base vendor? How commonly used is it? The two options are either to remove it from this list and allow it to download as a separate file when needed, or to do special handling in the RequireJS config in LMS so that it does not download separately (this is already done for most everything in this list, see build.js and require-config.js, searching for "underscore" for example).

clrux replied:
I've included Modernizr as a globally loaded script, because it's necessary for the AFontGarde icon font solution. Since icon fonts could (and probably are) used on every page, I felt it useful to make Modernizr globally available as well.

cahrens replied:
Then follow the same pattern as Underscore and JQuery. You will need to add a "defineDependency" line in require-config.js around line 48, and you will need to add the special 'empty:' mapping in build.js, around line 78.

cahrens commented:
@clrux Getting closer. The good news is that everything related to Draggabilly looks good. Bad news is that you have more work to do with the other libraries.

clrux commented:
@cahrens How's this? I don't really know what I'm looking for. CMS and LMS seem to be different...

cahrens commented:
@clrux You are looking for the modernizer file to not be downloaded as a separate file on your sandbox. Update your sandbox and take a look. You can then also remove this bit in the LMS requireJS config file:

 "modernizr": {
                 exports: "Modernizr"
             },

clrux commented:
@cahrens I believe I've figured it out. The reason `edx-icons` wasn't being cached was because it was being referenced in two different ways: one with a hyphen (`edx-icons`) and one without (`edxicons`). The override in `common.py` that provided the hash was not referencing the same file as RequireJS. They are synced up now and I don't see lose files. The sandboxes have been updated. Does this clear things up?

cahrens commented:
👍 🚀 🎆 😺 🍭 

(A side note-- I find dragging the captions to be a bit temperamental. If I drag too fast, they get "stuck" or move in a very jerky fashion. I've trained myself to drag slowly.)

clrux commented:
@cahrens :-) Thank you! Now... :beers: :beers: :beers: 

Yes, I've noticed that too.

cahrens commented:
As I'm sure you know, the Lettuce test failure is a known flaky (Diana is actually working on it right now!). There is no need for you to rerun Lettuce tests.

clrux commented:
I am going to merge this PR. The lettuce tests here are known to be flaky. Diana is working on them.
